firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -103,4 +125,129 @@ String getApplicationId() {
   String getName() {
     return firebaseApp.getName();
   }
+
+  /**
+   * Get firebase installation id for the {@link FirebaseApp}. Either from the persisted fid entry
+   * if already exists and is not error state or create a new Fid on the FIS Servers and persist in
+   * the shared prefs.
+   *
+   * <pre>
+   *     The workflow is:
+   *          Persisted Fid does not exists Or Persisted Fid is in Error state
+   *                                 |
+   *             return a newly created random Fid and persist the Fid entry
+   *
+   *
+   *
+   *                    Persisted Fid exists And is Registered
+   *                                  |
+   *                          return the Persisted Fid
+   *
+   *
+   *
+   *                  Persisted Fid exists and is in Unregistered state
+   *                                     |
+   *                Register the Fid on FIS Server and update the PersistedFidEntry
+   *                                     |
+   *                            return the Persisted Fid
+   *
+   * </pre>
+   */
+  @WorkerThread
+  private String getFirebaseInstallationId() throws FirebaseInstallationsException {
+
+    PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
+
+    if (persistedFidExistsAndRegistered(persistedFidEntry)) {
+      return persistedFidEntry.getFirebaseInstallationId();
+    }
+
+    if (persistedFidMissingOrInErrorState(persistedFidEntry)) {
+      return createFidAndPersistFidEntry();
+    }
+
+    if (persistedFidExistsAndUnregistered(persistedFidEntry)) {
+      Tasks.call(executor, () -> registerAndSaveFid(persistedFidEntry.getFirebaseInstallationId()));
+      return persistedFidEntry.getFirebaseInstallationId();
+    }
+
+    return null;
+  }
+
+  private static boolean persistedFidMissingOrInErrorState(PersistedFidEntry persistedFidEntry) {
+    return persistedFidEntry == null
+        || persistedFidEntry.getPersistedStatus() == PersistedFid.PersistedStatus.REGISTER_ERROR;
+  }
+
+  private static boolean persistedFidExistsAndRegistered(PersistedFidEntry persistedFidEntry) {
+    return persistedFidEntry != null
+        && persistedFidEntry.getPersistedStatus() == PersistedFid.PersistedStatus.REGISTERED;
+  }
+
+  private static boolean persistedFidExistsAndUnregistered(PersistedFidEntry persistedFidEntry) {
```

Comment:
I think, this method name is fine, but how about one of these?

#fidExistsButUnregistered
#fidExistsButNeedsRegistration
#fidExistsButNeedsRegistrationWithFis